CMakeLists.txt

```diff
@@ -108,10 +108,24 @@ dgl_feature_option(
     "dev"
 )
 
-# Set debug compile option for gdb, only happens when -DCMAKE_BUILD_TYPE=DEBUG
-if (NOT MSVC)
-  set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -DDEBUG -O0 -g3 -ggdb")
-endif(NOT MSVC)
+# Set optimization options for different build types.
+if (${BUILD_TYPE} STREQUAL "dev")
+  if (MSVC)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /Od")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Od")
+  else()
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O0 -g3 -ggdb")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -g3 -ggdb")
+  endif()
+else()
+  if (MSVC)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /O2")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /O2")
+  else()
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -O3")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O3")
```

Comment:
Shall we use -O2 or -O3 here?

Reply:
`/O2` is the most optimized on windows but `-O3` is the most optimized in gcc

Reply:
https://wiki.gentoo.org/wiki/GCC_optimization/en

Reply:
-O3: enables -O2 as well as optimizations that are expensive in terms of compile time and memory usage. Compiling with -O3 is not a guaranteed way to improve performance, and in fact, in many cases, can slow down a system due to larger binaries and increased memory usage. -O3 is also known to break several packages. Using -O3 is not recommended. However, it also enables -ftree-vectorize so that loops in the code get vectorized and will use AVX YMM registers.

Reply:
Unless we have strong reasons, let's stay on O2.

Reply:
source? I didn't see it in https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html

Reply:
let's use `-O2` then.